File: tobas_setup_assistant/src/tobas_setup_assistant/setting_widgets/propulsion_system/propeller.py
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ...setup_assistant import SetupAssistant
    from ...parameter_getters import ParamGetterWidget

import math
from overrides import override
from PyQt5.QtWidgets import QVBoxLayout

from ...parameter_getters import ParamGetterWidget_SpinBox
from .base import BaseSelectedLinkSettingWidget


class PropellerWidget(BaseSelectedLinkSettingWidget):
    NAME = "Propeller"

    def __init__(self, main: SetupAssistant, link_name: str) -> None:
        super().__init__(main, link_name)

        rows = QVBoxLayout()
        self.setLayout(rows)

        self._param_rows = QVBoxLayout()
        rows.addLayout(self._param_rows)

        num_blade_description = "Number of blades per propeller."
        self._num_blade = ParamGetterWidget_SpinBox("Number of blades", num_blade_description, minimum=1, default=2)
        self._param_rows.addWidget(self._num_blade)

        diameter_description = "Diameter of the propeller's rotational plane."
        self._diameter = ParamGetterWidget_SpinBox(
            "Propeller Diameter", diameter_description, minimum=1, default=10, suffix=" inch"
        )
        self._param_rows.addWidget(self._diameter)

        blade_chord_description = "Chord length at 75% of the distance from the blade's center."
        self._blade_chord = ParamGetterWidget_SpinBox(
            "75% Blade chord", blade_chord_description, minimum=1, default=15, suffix=" mm"
        )
        self._param_rows.addWidget(self._blade_chord)

        pitch_angle_description = "Twist angle at 75% of the distance from the blade's center."
        self._pitch_angle = ParamGetterWidget_SpinBox(
            "75% Blade pitch angle", pitch_angle_description, minimum=1, maximum=90, default=15, suffix=" deg"
        )
        self._param_rows.addWidget(self._pitch_angle)

        rows.addStretch()

    @override
    def is_valid(self) -> bool:
        return True

    @override
    def copy_from(self, src: PropellerWidget) -> None:
        for i in range(self._param_rows.count()):
            param_des: ParamGetterWidget = self._param_rows.itemAt(i).widget()
            param_src: ParamGetterWidget = src._param_rows.itemAt(i).widget()
            param_des.set(param_src.get())

    @override
    def dump_settings(self) -> dict:
        res = dict()
        for i in range(self._param_rows.count()):
            param: ParamGetterWidget = self._param_rows.itemAt(i).widget()
            res[param.name()] = param.get()
        return res

    @override
    def load_settings(self, data: dict) -> None:
        for i in range(self._param_rows.count()):
            param: ParamGetterWidget = self._param_rows.itemAt(i).widget()
            param.set(data[param.name()])

    def num_blade(self) -> int:
        return self._num_blade.get()

    def propeller_diameter(self) -> float:
        """[m]"""
        return self._diameter.get() * 0.0254

    def propeller_radius(self) -> float:
        """[m]"""
        return self.propeller_diameter() / 2

    def blade_chord(self) -> float:
        """[m]"""
        return self._blade_chord.get() / 1000

    def pitch_angle(self) -> float:
        """[rad]"""
        return math.radians(self._pitch_angle.get())

File: tobas_gui/tobas_setup_assistant/src/propulsion_system/propeller.cpp
#include <QVBoxLayout>

#include <tobas_std_tools/unit_conversions.hpp>

#include "tobas_setup_assistant/setting_tabs/propulsion_system/propeller.hpp"

namespace gui
{
namespace setup_assistant
{
PropellerWidget::PropellerWidget()
{
  auto rows = new QVBoxLayout(this);

  num_blade_ = new ParamGetterWidget_SpinBox("Number of blades", "Number of blades per propeller.");
  num_blade_->setMinimum(1);
  num_blade_->setValue(2);
  rows->addWidget(num_blade_);

  diameter_ = new ParamGetterWidget_SpinBox("Propeller Diameter", "Diameter of the propeller's rotational plane.");
  diameter_->setMinimum(1);
  diameter_->setValue(10);
  diameter_->setSuffix(" inch");
  rows->addWidget(diameter_);

  blade_chord_ =
    new ParamGetterWidget_SpinBox("75% Blade chord", "Chord length at 75% of the distance from the blade's center.");
  blade_chord_->setMinimum(1);
  blade_chord_->setValue(15);
  blade_chord_->setSuffix(" mm");
  rows->addWidget(blade_chord_);

  pitch_angle_ = new ParamGetterWidget_SpinBox(
    "75% Blade pitch angle", "Twist angle at 75% of the distance from the blade's center.");
  pitch_angle_->setMinimum(1);
  pitch_angle_->setMaximum(90);
  pitch_angle_->setValue(15);
  pitch_angle_->setSuffix(" deg");
  rows->addWidget(pitch_angle_);

  rows->addStretch();
}

const char* PropellerWidget::name()
{
  return "Propeller";
}

bool PropellerWidget::isValid()
{
  return true;
}

void PropellerWidget::copyFrom(const PropellerWidget* src)
{
  num_blade_->setValue(src->num_blade_->getValue());
  diameter_->setValue(src->diameter_->getValue());
  blade_chord_->setValue(src->blade_chord_->getValue());
  pitch_angle_->setValue(src->pitch_angle_->getValue());
}

YAML::Node PropellerWidget::dump()
{
  YAML::Node node(YAML::NodeType::Map);

  node[num_blade_->name()] = num_blade_->getValue();
  node[diameter_->name()] = diameter_->getValue();
  node[blade_chord_->name()] = blade_chord_->getValue();
  node[pitch_angle_->name()] = pitch_angle_->getValue();

  return node;
}

void PropellerWidget::load(const YAML::Node& node)
{
  num_blade_->setValue(node[num_blade_->name()].as<int>());
  diameter_->setValue(node[diameter_->name()].as<int>());
  blade_chord_->setValue(node[blade_chord_->name()].as<int>());
  pitch_angle_->setValue(node[pitch_angle_->name()].as<int>());
}

int PropellerWidget::numBlade() const
{
  return num_blade_->getValue();
}

double PropellerWidget::diameter() const
{
  return tobas_std::inch2meter(diameter_->getValue());
}

double PropellerWidget::radius() const
{
  return diameter() / 2;
}

double PropellerWidget::bladeChord() const
{
  return blade_chord_->getValue() * 1e-3;
}

double PropellerWidget::pitchAngle() const
{
  return tobas_std::deg2rad(pitch_angle_->getValue());
}
}  // namespace setup_assistant
}  // namespace gui
